### PR TITLE
Add groupby functions for dates and timestamps

### PR DESCRIPTION
**Pull request by ssanjay1**
```
// syntax for seconds, minutes, hours, days, week, month, quarter, year
client(BoundariesUsState).aggregateOrThrow({
  select: {...},
  groupBy: { foundedDate: { duration: [10,"days"] } },
});

// Ranges
client(BoundariesUsState).aggregateOrThrow({
  select: {...},
  groupBy: {
    population: {
      ranges: [{
        startVal: "2024-03-07T19:26:30Z",
        endVal: "2024-03-08T19:26:30Z",
      }, {
        startVal: "2024-03-10T19:26:30Z,
        endVal: "2024-03-11T19:26:30Z",
      }],
    },
  },
});
```

### CODE REVIEW

File: packages/client/src/query/aggregations/GroupByClause.ts [hunk 1]
```diff
@@ -25,11 +25,52 @@ export type GroupByClause<
...
+  | "SECONDS"
+  | "MINUTES"
+  | "HOURS"
+  | "DAYS"
+  | "WEEKS"
+  | "MONTHS"
+  | "YEARS"
+  | "QUARTERS";
+
+export const TimeUnitMapping = {
```
Comment (ssanjay1) [inline packages/client/src/query/aggregations/GroupByClause.ts:50]: Was trying to emulate the flexibility we gave here https://github.com/palantir/osdk-ts/blob/012aae96dcbb063e8d6ca5f1efb4e74985da97d5/packages/client/src/query/WhereClause.ts#L63C1-L80C20 for the within clause, but may have added way too much, can definitely cut down

File: examples-extra/basic/sdk/src/generatedNoCheck/Ontology.ts [hunk 2]
```diff
@@ -27,7 +28,14 @@ const _Ontology = {
...
 } satisfies OntologyDefinition<
-  'WeatherStation' | 'BoundariesUsState' | 'Todo' | 'Person' | 'Employee' | 'Venture' | 'ObjectTypeWithAllPropertyTypes'
+  | 'WeatherStation'
+  | 'BoundariesUsState'
+  | 'BuilderDeploymentState'
+  | 'Todo'
+  | 'Person'
+  | 'Employee'
+  | 'Venture'
+  | 'ObjectTypeWithAllPropertyTypes'
```
Comment (mfedderly) [inline examples-extra/basic/sdk/src/generatedNoCheck/Ontology.ts:38]: This is probably a semantic merge conflict with https://github.com/palantir/osdk-ts/pull/124, I'll resolve later.

File: packages/client/src/query/aggregations/GroupByClause.ts [hunk 3]
```diff
@@ -25,11 +25,52 @@ export type GroupByClause<
 };
 export type StringGroupByValue = "exact" | { exactWithLimit: number };
 
-export type GroupByRange = [number, number];
+export type GroupByRange<T> = [T, T];
+export type DurationGroupBy = [number, keyof typeof TimeUnitMapping];
```
Comment (mfedderly) [inline packages/client/src/query/aggregations/GroupByClause.ts:29]: Does this work? Or can I specify any string as a duration time unit? (combined with the satisfies `Record<string, TimeUnit>` below)

File: packages/client/src/query/aggregations/GroupByMapper.ts [hunk 4]
```diff
@@ -27,4 +28,6 @@ export interface GroupByMapper {
   byte: NumericGroupByValue;
   double: NumericGroupByValue;
   integer: NumericGroupByValue;
+  timestamp: TimestampGroupByValue;
+  datetime: TimestampGroupByValue;
```
Comment (mfedderly) [inline packages/client/src/query/aggregations/GroupByMapper.ts:32]: Can we do this? Or does datetime need its own definition that doesn't let you do anything smaller than a day?

File: packages/client/src/query/aggregations/GroupByClause.ts (hunk 3 above)
Comment (mfedderly) [reply packages/client/src/query/aggregations/GroupByClause.ts:29]: Additionally it sounds like "YEARS" "QUARTERS" "MONTHS" "WEEKS" all require value to be 1 (see AggregationDurationGroupingV2's jsdoc), which isn't captured in our types.

File: examples-extra/basic/sdk/src/generatedNoCheck/Ontology.ts (hunk 2 above)
Comment (mfedderly) [reply examples-extra/basic/sdk/src/generatedNoCheck/Ontology.ts:38]: Actually I'm going to merge mine, can you eventually rebase and just run codegen before you merge?

File: packages/client/src/query/aggregations/GroupByMapper.ts (hunk 4 above)
Comment (ssanjay1) [reply packages/client/src/query/aggregations/GroupByMapper.ts:32]: Based on the backend code, date also seems to accept things smaller than day, but I'll double check with backend team

File: examples-extra/basic/sdk/src/generatedNoCheck/Ontology.ts (hunk 2 above)
Comment (ssanjay1) [reply examples-extra/basic/sdk/src/generatedNoCheck/Ontology.ts:38]: Yea I can rebase, go for it

File: packages/client/src/query/aggregations/GroupByClause.ts (hunk 3 above)
Comment (ssanjay1) [reply packages/client/src/query/aggregations/GroupByClause.ts:29]: Let me check on the type again, can write a type test for it too just to make sure. And ah thanks for the flag on the years, quarters etc.

File: packages/client/src/query/aggregations/GroupByMapper.ts [hunk 5]
```diff
@@ -27,4 +29,6 @@ export interface GroupByMapper {
   byte: NumericGroupByValue;
   double: NumericGroupByValue;
   integer: NumericGroupByValue;
+  timestamp: TimestampGroupByValue;
+  datetime: DateGroupByValue;
```
Comment (ssanjay1) [inline packages/client/src/query/aggregations/GroupByMapper.ts:33]: So dates can technically accept groupby durations smaller than a day, but they give back very weird results (Gateway should probably throw errors on these). For now, just restricting it on our end

**Review by mfedderly — APPROVED** [review]
🎉

**Pull request merged**